File: Source/Utility/Hooks/use-debounce.hook.ts
import { BindingOrValue, getBindingValue } from "@rbxts/pretty-roact-hooks";
import { useBinding } from "@rbxts/roact-hooked";

export default function useDebounce(Delay: BindingOrValue<number>) {
    const [ Last, Update ] = useBinding(0);

    const Can = (Now: number = os.clock()): boolean => {
        return (getBindingValue(Last) + getBindingValue(Delay)) <= Now;
    };

    const Try = (Callback?: () => void): boolean => {
        const Now = os.clock();

        if (!Can(Now)) return false;
        if (Callback) task.defer(Callback);

        Update(Now);
        return true;
    };

    return $tuple(Can, Try);
};

File: Output/Utility/Hooks/use-debounce.hook.lua
-- Compiled with roblox-ts v2.2.0
local TS = _G[script]
local getBindingValue = TS.import(script, TS.getModule(script, "@rbxts", "pretty-roact-hooks").out).getBindingValue
local useBinding = TS.import(script, TS.getModule(script, "@rbxts", "roact-hooked").src).useBinding
local function useDebounce(Delay)
	local Last, Update = useBinding(0)
	local Can = function(Now)
		if Now == nil then
			Now = os.clock()
		end
		return (getBindingValue(Last) + getBindingValue(Delay)) <= Now
	end
	local Try = function(Callback)
		local Now = os.clock()
		if not Can(Now) then
			return false
		end
		if Callback then
			task.defer(Callback)
		end
		Update(Now)
		return true
	end
	return Can, Try
end
return {
	useDebounce = useDebounce,
}
